p2p/src/publisher.rs

```diff
@@ -35,10 +39,77 @@ const OPTIMISM: f64 = 0.5;
 
 // Manages publishing provider records regularly over an interval.
 pub struct Publisher {
+    start_providing_results_tx: Sender<AddProviderResult>,
+    batches_rx: Receiver<Vec<Key>>,
+    metrics: Metrics,
+}
+
+impl Publisher {
+    pub fn new(interval: Duration, block_store: SQLiteBlockStore, metrics: Metrics) -> Self {
+        // Channel for result of each start_provide query.
+        let (results_tx, results_rx) = channel(MAX_RUNNING_QUERIES * 2);
+
+        // We should rarely be behind by more than a single batch.
+        // If we are backpressure is good as we cannot use the new batch.
+        let (batches_tx, batches_rx) = channel(1);
+
+        // Do real work of the publisher on its own task.
+        let mut stream = PublisherStream::new(results_rx, interval, block_store, metrics.clone());
+        let task_metrics = metrics.clone();
+        tokio::spawn(async move {
+            while let Some(batch) = stream.next().await {
```

Comment:
So:
- The `Publisher` drives `PublisherStream` in a separate task.
- `PublisherStream` builds and returns batches into a channel that gets polled by the swam task.
- The swarm publishes a batch and returns results via another channel that gets consumed first before a new batch is returned.

Does this sound like an accurate description?

Reply:
And it's the SQL lookups/batching logic that's been moved out from the swarm task to its own task, via `PublisherStream`.

Reply:
>PublisherStream builds and returns batches into a channel that gets polled by the swarm task.

This part is technically incorrect. Both the Publisher and PublisherStream types implement the Stream trait. The swarm is unware that PublisherStream exists or that the channel exists. So the swarm task polls the Publisher for the next batch. The only work the Publisher does is to read a channel if something is available. 

> The swarm publishes a batch and returns results via another channel that gets consumed first before a new batch is returned.

Yes except the swarm is unware of the results channel as well. It simply calls the handle method on the Publisher which puts the result on a channel that the PublisherStream reads. 

So generally correct except that the Publisher abstraction means that the swarm doesn't care that channels and a separate task are being used under the hood.

Perhaps I should rename PublisherStream to PublisherWorker? That way its not misleading that they both implement Stream?

Reply:
Ah gotcha, makes sense, ty! I got confused by the two layers of streams.

Yeah, I think calling it PublisherWorker would help.